File: _22_Maximum_Depth_of_Binary_Tree.ts

class TreeNode4 {
    val: number;
    left: TreeNode4 | null;
    right: TreeNode4 | null;

    constructor(val?: number, left?: TreeNode4 | null, right?: TreeNode4 | null) {
        this.val = (val === undefined ? 0 : val);
        this.left = (left === undefined ? null : left);
        this.right = (right === undefined ? null : right);
    }
}

function maxDepth(root: TreeNode4 | null): number {
    if (root === null) return 0;

    const heightLeft = maxDepth(root.left);
    const heightRight = maxDepth(root.right);

    return 1 + Math.max(heightLeft, heightRight);
}


function main() {
    const root1 = new TreeNode4(1,
        new TreeNode4(2, new TreeNode4(3), new TreeNode4(4)),
        new TreeNode4(2, new TreeNode4(4), new TreeNode4(3))
    );
    console.log("Ejemplo 1:", maxDepth(root1)); 

    const root2 = new TreeNode4(1,
        new TreeNode4(2, null, new TreeNode4(3)),
        new TreeNode4(2, null, new TreeNode4(3))
    );
    console.log("Ejemplo 2:", maxDepth(root2));

    console.log("Ejemplo 3:", maxDepth(null)); 

    const root4 = new TreeNode4(1);
    console.log("Ejemplo 4:", maxDepth(root4)); 
}

main();

File: _22_Maximum_Depth_of_Binary_Tree.java
public class _22_Maximum_Depth_of_Binary_Tree {
    public static void main(String[] args) {
        Solution sol = new Solution();

        TreeNode root1 = new TreeNode(1,
            new TreeNode(2, new TreeNode(3), new TreeNode(4)),
            new TreeNode(2, new TreeNode(4), new TreeNode(3))
        );
        System.out.println("Ejemplo 1: " + sol.maxDepth(root1)); 

        TreeNode root2 = new TreeNode(1,
            new TreeNode(2, null, new TreeNode(3)),
            new TreeNode(2, null, new TreeNode(3))
        );
        System.out.println("Ejemplo 2: " + sol.maxDepth(root2));
    }
}

class TreeNode {
    int val;
    TreeNode left, right;
    TreeNode() {}
    TreeNode(int val) { this.val = val; }
    TreeNode(int val, TreeNode left, TreeNode right) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}

//My solution
class Solution {
    public int maxDepth(TreeNode root) {

        if ( root == null )  return 0;

        int heightLeft = maxDepth( root.left );
        int heightRight = maxDepth( root.right );

        
        return 1 + Math.max(heightLeft, heightRight);
    }
}